Fixes #22952 - Handle SUSE errata's epoch dates

johnpmitsch opened the pull request:
SUSE errata dates are stored in pulp as seconds since epoch dates,
whereas Red Hat errata dates are formatted date strings. Since pulp
stores these "as-is" in a string, we need to convert the epoch date
to a formatted date string in the indexing step, so the epoch date
can be stored in our "issued" and "updated" database fields for
Erratum, which are date type fields.

jlsherrill commented on app/models/katello/erratum.rb:
```diff
@@ -118,6 +118,12 @@ def update_from_json(json)
+      if custom_json["issued"].to_i.to_s == custom_json["issued"]
+        custom_json["issued"] = epoch_to_datetime(custom_json["issued"])
+        custom_json["updated"] = epoch_to_datetime(custom_json["updated"]) unless custom_json["updated"].blank?
+      end
```
mind adding a test for this?

jlsherrill commented on app/models/katello/erratum.rb:
```diff
@@ -118,6 +118,12 @@ def update_from_json(json)
+      # handle SUSE epoch dates
+      if custom_json["issued"].to_i.to_s == custom_json["issued"]
+        custom_json["issued"] = epoch_to_datetime(custom_json["issued"])
+        custom_json["updated"] = epoch_to_datetime(custom_json["updated"]) unless custom_json["updated"].blank?
```
I would probably have different checks for issued and updated.  Sure most likely if its an epoch with issued, its going to be an epoch with updated but that may not always be true, so i would handle them separately.

jlsherrill commented:
[test katello]

johnpmitsch commented:
[test katello]

johnpmitsch commented:
[test katello]

johnpmitsch commented:
[test katello]

johnpmitsch commented:
@jlsherrill I updated the tests since a lot of them were using date objects in the values in the json passed  to `update_from_json` as opposed to strings, which was causing failures with the new methods.  Also the "json" is really a hash :confused:

jlsherrill commented:
@johnpmitsch could you just call '.as_json' on the errata objects?  that seems much simpler

johnpmitsch commented:
@jlsherrill that is what I tried first, but `update_from_json` actually does expect a hash. Since that hash is originally derived from json, all the values are strings.

johnpmitsch commented:
@jlsherrill the "json" here is a Hash that originally came from json, where the values are strings, i.e.  "issued"=>"2013-01-27 16:08:09".
the lines I added are assuming this for the date fields.
```
From: /home/vagrant/katello/app/models/katello/erratum.rb @ line 120 Katello::Erratum#update_from_json:

    117: def update_from_json(json)
    118:   keys = %w(title id severity issued type description reboot_suggested solution updated summary)
    119:   binding.pry
 => 120:   custom_json = json.slice(*keys)
    121:
    122:   if self.updated.blank? || (custom_json['updated'].to_datetime != self.updated.to_datetime)
    123:     custom_json['errata_id'] = custom_json.delete('id')
    124:     custom_json['errata_type'] = custom_json.delete('type')
    125:     custom_json['updated'] = custom_json['updated'].blank? ? custom_json['issued'] : custom_json['updated']
    126:     self.update_attributes!(custom_json)
    127:
    128:     unless json['references'].blank?
    129:       update_bugzillas(json['references'].select { |r| r['type'] == 'bugzilla' })
    130:       update_cves(json['references'].select { |r| r['type'] == 'cve' })
    131:     end
    132:   end
    133:   update_packages(json['pkglist']) unless json['pkglist'].blank?
    134: end

[1] pry(#<Katello::Erratum>)> json.class
=> ActiveSupport::HashWithIndifferentAccess	
[2] pry(#<Katello::Erratum>)> json
=> {"repository_memberships"=>["9301b3e7-a9cb-4adb-b532-8866bdf3b101"],
 "_href"=>"/pulp/api/v2/content/units/erratum/b692f509-9045-4ee8-b3c9-87fa050b2856/",
 "issued"=>"2013-01-27 16:08:09",
 "references"=>[],
 "pulp_user_metadata"=>{},
 "_content_type_id"=>"erratum",
 "id"=>"RHEA-2012:0058",
 "from"=>"errata@redhat.com",
 "severity"=>"",
 "title"=>"Gorilla_Erratum",
 "children"=>{},
 "version"=>"1",
 "reboot_suggested"=>false,
 "type"=>"enhancement",
 "pkglist"=>
  [{"packages"=>
     [{"src"=>"http://www.fedoraproject.org", "name"=>"gorilla", "sum"=>nil, "filename"=>"gorilla-0.62-1.noarch.rpm", "epoch"=>nil, "version"=>"0.62", "release"=>"1", "arch"=>"noarch"}],
    "name"=>"1",
    "short"=>""}],
 "status"=>"stable",
 "updated"=>"",
 "description"=>"Gorilla_Erratum",
 "_last_updated"=>"2018-04-09T15:12:43Z",
 "pushcount"=>"",
 "rights"=>"",
 "solution"=>"",
 "summary"=>"",
 "release"=>"1",
 "_id"=>"b692f509-9045-4ee8-b3c9-87fa050b2856"} 
 ```

 Our tests use the errata fixtures which use Time objects https://github.com/johnpmitsch/katello/blob/master/test/fixtures/models/katello_errata.yml#L9
https://github.com/johnpmitsch/katello/blob/e644d56f76ede125607e52f855ad9953705489d0/test/models/erratum_test.rb#L152

passing in a Hash with Time objects into update_from_json, which causes errors in the existing tests because of the the new date converting lines in update_from_json

jlsherrill commented on test/models/erratum_test.rb:
```diff
@@ -90,7 +98,7 @@ def test_not_applicable_to_hosts
       errata = katello_errata(:security)
       json = errata.attributes.merge('description' => 'an update', 'updated' => Time.now, 'reboot_suggested' => true)
-      errata.update_from_json(json)
+      errata.update_from_json(errata_json(json))
```
i think just calling  `errata.update_from_json(json.as_json)`
would work here.  It seemed to work fine for me locally.

johnpmitsch commented:
@jlsherrill the new test format ok with you?
